pkg/core/tekton/tekton.go

```diff
@@ -0,0 +1,405 @@
+package tekton
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"github.com/tektoncd/triggers/pkg/apis/triggers/v1alpha1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"knative.dev/pkg/apis"
+
+	"github.com/fuseml/fuseml-core/gen/codeset"
+	"github.com/fuseml/fuseml-core/gen/workflow"
+	"github.com/fuseml/fuseml-core/pkg/core/tekton/builder"
+)
+
+// WorkflowBackend implements the FuseML WorkflowBackend interface for tekton
+type WorkflowBackend struct {
+	namespace     string
+	tektonClients *clients
+}
+
+// NewWorkflowBackend initializes Tekton backend
+func NewWorkflowBackend(namespace string) *WorkflowBackend {
+	return &WorkflowBackend{namespace, newClients(namespace)}
+}
+
+// CreatePipeline receives a FuseML workflow and creates a Tekton pipeline from it
+func (t *WorkflowBackend) CreatePipeline(ctx context.Context, logger *log.Logger, workflow *workflow.Workflow) error {
+	pipeline := generatePipeline(*workflow, t.namespace)
+	logger.Printf("Creating tekton pipeline for workflow: %s...", workflow.Name)
+	_, err := t.tektonClients.PipelineClient.Create(ctx, pipeline, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RunPipeline creates a PipelineRun with its default values for received workflow and codeset
+func (t *WorkflowBackend) RunPipeline(ctx context.Context, workflowName string, codeset codeset.Codeset) error {
+	pipeline, err := t.tektonClients.PipelineClient.Get(ctx, workflowName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	pipelineRun, err := generatePipelineRun(pipeline, codeset)
+	if err != nil {
+		return err
+	}
+
+	_, err = t.tektonClients.PipelineRunClient.Create(ctx, pipelineRun, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// CreateListener creates tekton resources required to have a listener ready for triggering the pipeline
+func (t *WorkflowBackend) CreateListener(ctx context.Context, logger *log.Logger, workflowName string) (string, error) {
+
+	pipeline, err := t.tektonClients.PipelineClient.Get(ctx, workflowName, metav1.GetOptions{})
+	if err != nil {
+		return "", err
```

Comment:
Wouldn't make sense to log the error at the top level, for example at `workflow.go`?

Reply:
Look at it this way: the farther away from the error source you're logging the error, the more difficult it will be to pinpoint where in the code the error occurred for someone looking to debug it.

Reply:
This may be a personal grievance I had with some of the code I had to debug. I often had to "guess" where the error actually occurred because instead of being logged close to the source, it was passed up the call chain to a point that had nothing to do with the error.

Reply:
Maybe wrapping the error with `fmt.Errorf` helps with that?

Reply:
I have made the changes using `fmnt.Errorf` to wrap the errors which are then logged at `workflow.go`. Let me know if you think that is not enough.